server/scsynth/SC_Bela.cpp

```diff
@@ -0,0 +1,513 @@
+            int32 bufCounter = mWorld->mBufCounter;
+            int32* tch;
+
+            // copy+touch inputs
+            tch = inTouched;
+            memcpy(inBuses, belaContext->audioIn, sizeof(belaContext->audioIn[0]) * bufFrames * minInputs);
+            for (int k = 0; k < minInputs; ++k) {
+                *tch++ = bufCounter;
+            }
+
+            memcpy(inBuses + minInputs * bufFrames, belaContext->analogIn,
+                   sizeof(belaContext->analogIn[0]) * bufFrames * anaInputs);
+            for (int k = minInputs; k < (minInputs + anaInputs); ++k) {
+                *tch++ = bufCounter;
+            }
+
+            // run engine
+            int64 schedTime;
+            int64 nextTime = oscTime + oscInc;
+
+            while ((schedTime = mScheduler.NextTime()) <= nextTime) {
+                float diffTime = (float)(schedTime - oscTime) * oscToSamples + 0.5;
+                float diffTimeFloor = floor(diffTime);
+                world->mSampleOffset = (int)diffTimeFloor;
+                world->mSubsampleOffset = diffTime - diffTimeFloor;
+
+                if (world->mSampleOffset < 0)
+                    world->mSampleOffset = 0;
+                else if (world->mSampleOffset >= world->mBufLength)
+                    world->mSampleOffset = world->mBufLength - 1;
+
+                SC_ScheduledEvent event = mScheduler.Remove();
+                event.Perform();
+            }
+
+            world->mSampleOffset = 0;
+            world->mSubsampleOffset = 0.f;
+            World_Run(world);
+
+            // copy touched outputs
+            tch = outTouched;
+
+            for (int k = 0; k < minOutputs; ++k) {
+                if (*tch++ == bufCounter) {
+                    memcpy(belaContext->audioOut + k * bufFrames, outBuses + k * bufFrames,
+                           sizeof(belaContext->audioOut[0]) * bufFrames);
+                }
+            }
+
+            for (int k = minOutputs; k < (minOutputs + anaOutputs); ++k) {
+                if (*tch++ == bufCounter) {
+                    unsigned int analogChannel = k - minOutputs; // starting at 0
+                    memcpy(belaContext->analogOut + analogChannel * bufFrames, outBuses + k * bufFrames,
+                           sizeof(belaContext->analogOut[0]) * bufFrames);
+                }
+            }
+
+            // advance OSC time
+            mOSCbuftime = oscTime = nextTime;
+        }
+
+        if (mWorld->mBelaScope)
+            mWorld->mBelaScope->logBuffer();
+
+    } catch (std::exception& exc) {
+        scprintf("SC_BelaDriver: exception in real time: %s\n", exc.what());
+    } catch (...) {
+        scprintf("SC_BelaDriver: unknown exception in real time\n");
+    }
+
+    mAudioSync.Signal();
+}
+
+// ====================================================================
+
+// HW_DETECT_HACK: these functionalities should instead be exposed by libbela
+typedef struct _BelaHwConfig // HW_DETECT_HACK
```

Comment:
is there any plan to remedy this in libbela? i can submit a PR to bela if that would help.

Reply:
hmmm ... I think I could add something like that, while hiding some fields for internal use only.

Reply:
i.e." the `activeCodec` and `disabledCodec` should really be just private implementation details.

Reply:
I will add it tomorrow

Reply:
https://github.com/BelaPlatform/Bela/commit/2c091287784d40ad8fc3259561c84783d1b57f04

and

36015f35b

though the version check in the latter should probably be moved to `FindBela.cmake`?

Reply:
thanks!!

> though the version check in the latter should probably be moved to FindBela.cmake?

i don't know how to properly do version checks for CMake find-modules. there seem to be instructions on how to do it in the CMake documentation [here](https://cmake.org/cmake/help/latest/command/find_package.html#version-selection) but i haven't taken the time to really understand what is being said there yet. i also own a copy of Professional CMake which is usually illuminating regarding CMake best practices, so i can check there, too. it would be nice to adopt whatever version checking method is considered best by CMake experts, as this could turn out to be useful for other projects. since this is a somewhat open-ended endeavor i'm inclined for now to leave this check as-is, since it certainly works, and add this to the list of post-first-PR issues to tackle.